### ISSUE TITLE
Error: Can't apply this patch as its origin declaration isn't external.

### ISSUE THREAD

**Issue by grundid** [dart-lang/sdk#42475]
When switching from dart 2.7.2 to 2.8.1 I'm seeing the following build error:

Dart2Js finished with:

/Downloads/dart-sdk-2.8.1/lib/_internal/js_runtime/lib/mirrors_patch_cfe.dart:30:31:
Error: Can't apply this patch as its origin declaration isn't external.
  Map<Uri, LibraryMirror> get libraries;
                              ^
/Downloads/dart-sdk-2.8.1/lib/mirrors/mirrors.dart:88:31:
Info: This is the origin declaration.
  Map<Uri, LibraryMirror> get libraries;
                              ^
/Downloads/dart-sdk-2.8.1/lib/_internal/js_runtime/lib/mirrors_patch_cfe.dart:31:21:
Error: Can't apply this patch as its origin declaration isn't external.
  IsolateMirror get isolate;
                    ^
/Downloads/dart-sdk-2.8.1/lib/mirrors/mirrors.dart:102:21:
Info: This is the origin declaration.
  IsolateMirror get isolate;
                    ^
/Downloads/dart-sdk-2.8.1/lib/_internal/js_runtime/lib/mirrors_patch_cfe.dart:32:18:
Error: Can't apply this patch as its origin declaration isn't external.
  TypeMirror get dynamicType;
                 ^
/Downloads/dart-sdk-2.8.1/lib/mirrors/mirrors.dart:107:18:
Info: This is the origin declaration.
  TypeMirror get dynamicType;
                 ^
/Downloads/dart-sdk-2.8.1/lib/_internal/js_runtime/lib/mirrors_patch_cfe.dart:33:18:
Error: Can't apply this patch as its origin declaration isn't external.
  TypeMirror get voidType;
                 ^
/Downloads/dart-sdk-2.8.1/lib/mirrors/mirrors.dart:112:18:
Info: This is the origin declaration.
  TypeMirror get voidType;
                 ^
/Downloads/dart-sdk-2.8.1/lib/_internal/js_runtime/lib/mirrors_patch_cfe.dart:34:18:
Error: Can't apply this patch as its origin declaration isn't external.
  TypeMirror get neverType;
                 ^
/Downloads/dart-sdk-2.8.1/lib/mirrors/mirrors.dart:118:18:
Info: This is the origin declaration.
  TypeMirror get neverType;
                 ^
Error: Compilation failed.

**Comment by sigmundch** [comment]
Thanks for reporting this issue.

@johnniwinther did the CFE add some additional checks like these recently (in the 2.8 timeframe, but post 2.7?). Is the proper fix to remove the duplicate getter definitions?

I'm surprised that this was not detected in our build process. Turns out that dart2js_target doesn't include `dart:mirrors` by default, so it is being compiled on the client side.

@grundid we'll work on getting a fix ready for this, but note that this is only triggered if you have imports to `dart:mirrors`. The mirrors library is not supported on web clients, and all its APIs will throw, so if it is at all possible to remove the import, that should address this on your end as a workaround.

**Comment by sigmundch** [comment]
https://dart-review.googlesource.com/c/sdk/+/152406 might be enough to fix this.

**Comment by sigmundch** [comment]
/cc @nshahan 

I now remember that earlier this year, in 35cd61819a51ded1015e4da7cc31ddc04e65b241, we removed this on purpose and that the intent was that we'd issue a compile-time error indicating that `dart:mirrors` was not supported anymore.  We reverted the breaking change and added dart:mirrors back in 344e149a89bd2fc7dd73d538b6fbef8db8192a8d, but we didn't add it by default on the pre-compiled SDK, and as a result the new CFE checks were no longer being checked when we build our SDK.

I'll clean up the change above to fix the immediate issue. However, we should look into adding the error we mention in the CHANGELOG in 344e149a89bd2fc7dd73d538b6fbef8db8192a8d (to produce the compile-time error when opting-in to the next set of breaking changes).

@grundid - I'm curious to hear about this dependency you have on dart:mirrors to better understand whether this will be affecting many other developers such as yourself.
